llama : std::move llm_bigram_bpe from work_queue

danbev opened the pull request:
This commit updates the retrieval of llm_bigram_bpe objects from work_queue.top() by using std::move.

The motivation for this is to avoid the copying of the std::string `text` member of the llm_bigram_bpe struct.



- [x] I have read the [contributing guidelines](https://github.com/ggerganov/llama.cpp/blob/master/CONTRIBUTING.md)
- Self-reported review complexity:
  - [x] Low
  - [ ] Medium
  - [ ] High

danbev committed:
llama : std::move llm_bigram_bpe from work_queue

This commit updates the retrieval of llm_bigram_bpe objects from
work_queue.top() by using std::move.

The motivation for this is to avoid the copying of the std::string
`text` member of the llm_bigram_bpe struct.

ggerganov commented on src/llama-vocab.cpp:
```diff
@@ -520,7 +520,7 @@ struct llm_tokenizer_bpe {
             // build token(s)
             while (!work_queue.empty()) {
-                auto bigram = work_queue.top();
+                auto bigram = std::move(const_cast<llm_bigram_bpe&>(work_queue.top()));
```
Prefer to avoid `const_cast`. Better to switch to a heap container that supports move semantic

danbev replied:
I've added a commit that tries to address this. Was this similar to what you had in mind?

danbev committed:
squash! llama : std::move llm_bigram_bpe from work_queue

Introduced a MovablePriorityQueue class to allow moving elements
out of the priority queue for llm_bigram_bpe.

ggerganov commented on src/llama-vocab.cpp:
```diff
@@ -321,6 +321,21 @@ struct llm_tokenizer_spm {
 // TODO: there are a lot of common parts between spm and bpe tokenizers, should be refactored and reused
 
+template<typename T, typename Container = std::vector<T>, typename Compare = std::less<typename Container::value_type>>
+class MovablePriorityQueue : public std::priority_queue<T, Container, Compare> {
```
Change the name to `snake_case`:

```suggestion
class llama_priority_queue : public std::priority_queue<T, Container, Compare> {
```

danbev committed:
squash! llama : std::move llm_bigram_bpe from work_queue

Rename MovablePriorityQueue to lama_priority_queue.

ggerganov commented on src/llama-vocab.cpp:
```diff
@@ -321,6 +321,21 @@ struct llm_tokenizer_spm {
 // TODO: there are a lot of common parts between spm and bpe tokenizers, should be refactored and reused
 
+template<typename T, typename Container = std::vector<T>, typename Compare = std::less<typename Container::value_type>>
+class lama_priority_queue : public std::priority_queue<T, Container, Compare> {
```
```suggestion
class llama_priority_queue : public std::priority_queue<T, Container, Compare> {
```

danbev replied:
Sorry about that, I'll update this shortly.

danbev committed:
squash! llama : std::move llm_bigram_bpe from work_queue

Rename lama_priority_queue -> llama_priority_queue.
